grafana/icons
# POC: Generate Icon components from SVGs

This PR repurposes this repo for the new icons package, with a POC demonstrating how this package might function. 

**How it works**
1. SVGs for the icons are uploaded to the `svg` folder.
2. `yarn icons:create` converts the SVGs into TypeScript React components inside the `src/icons` folder. Each component is wrapped in an `IconBase` component for easier customisation of the icons. A custom template in `templates/icon.js` is used to generate each component.
3. `yarn build` compiles the generated code and creates esm + cjs bundles in the `dist` folder. This is mostly a copy/paste of the same logic from `@grafana` packages.
4. [Not implemented] Once the package is published, the new icons become available. 

**How to test locally, within Grafana**

Due to issues with package linking in yarn version 3 and above, additional steps are required to enable it to work correctly:
1. Clone this repo and check out the `svg-to-component` branch.
2. Run `yarn install && yarn build`.
3. Inside the cloned repo, run `yarn link`.
4. Navigate to the Grafana folder.
5. Run `yarn link -p <relative-path-to-the-icons-package>`. In the package.json it'd create a new resolution: ```"resolutions": {
    "@grafana/icons": "portal:/absolute-path-to-the-package"
  }```
6. Add the package to dependencies and run `yarn install`: 
```
"devDependencies": {
  "@grafana/icons": "*",
}
```
7. Import the icons inside Grafana: `import { AngleRight } from "@grafana/icons";`. The new icon API is largely similar to the one of the current `Icon` component. 

**TODOs**

- [ ] CI for publishing the package and its new versions. The publish should happen on the package.json version increment. 
- [ ] A Playground to be able to preview the icons (or should it be done in the Sage website instead?)

**Questions**
- Do all the icons needs to be wrapped in `forwardRef`? We currently do it for the `Icon` component, but I wonder if it's used? It will add quite a bit of complexity to the icon template. 
- Do we need both cjs and esm builds? 

Fixes: https://github.com/grafana/grafana/issues/81682.
